Work in progress for frequency trucks

lisavancasteren opened the pull request:
The button 'buy an extra truck' is registered in the factorycontroller via the function frequencyTrucks. The function updates the number of trucks (maxSimultaneousOrder), after checking if there is enough money to do so (validate Trucks). Added config: costExtraTruck (now set at 100, should be changed probably).
Closes #73.

lisavancasteren committed:
Merge remote-tracking branch 'refs/remotes/LCfP/master'

lisavancasteren committed:
Button for frequency trucks

N-Wouda commented on src/views/template/factory/factory.html:
```diff
@@ -32,5 +32,16 @@
                 {{#l}}Place order{{/l}}
             </button>
         </form>
+        <form name="frequencyTrucks" action="">
```
This part defines two forms (should be only one).

I think a button group may be better, since we also need an "increase capacity" button or similar. This would make that sufficiently generalisable. Then, bind to the click event on the button, like `$("#buy-truck").click(function (e) {...});` etc.
```html
<div class="btn-group" role="group">
    <button id="buy-truck">Buy truck</button>
</div>
```

N-Wouda replied:
Be sure to update the HTML here!

N-Wouda commented on src/controllers/core/controller.js:
```diff
@@ -114,4 +114,9 @@ class Controller
         $("#money").html(GAME.model.config.money.toFixed(2));
     }
+
+    _updateMaxSimultaneousOrder(amount)
```
I don't think this is necessary. Can be done in Factorycontroller.

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -105,6 +118,22 @@ class FactoryController extends OrderController
+
+    validateTruck(truck)
+    {
+        if(truckCost>GAME.model.config.money) {
```
Spaces around operators, like `truckCost >= GAME.model.config.money`. Where is `truckCost` defined?

lisavancasteren replied:
In the config file? But it is not recognized as a variable, so I guess it should be somewhere else:)

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -105,6 +118,22 @@ class FactoryController extends OrderController
+    validateTruck(truck)
+    {
+        if(truckCost>GAME.model.config.money) {
+            toastr.error(Controller.l("You cannot afford an extra truck!"));
```
`"You cannot afford an extra truck!"` should be added to the language file.

N-Wouda commented:
Overall I'm pretty happy with the result. We may need to add a visual display of how many orders can be placed at any given time, but I am not sure that should be part of this PR.

I placed a few comments, and I'll have another look when those have been resolved!

N-Wouda commented:
@lisavancasteren I'll walk you through the merge conflict if necessary. For now, there's no need to worry about that!

lisavancasteren commented:
Thanks for the feedback Niels, I will look at it!

Op 4 jan. 2017 20:33 schreef "Niels Wouda" <notifications@github.com>:

> @lisavancasteren <https://github.com/lisavancasteren> I'll walk you
> through the merge conflict if necessary. For now, there's no need to worry
> about that!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LCfP/Serious-Game-for-Kids/pull/93#issuecomment-270464382>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV93kWRIpVIYwkTRdXD3BbTMDDT1kCPYks5rO_QEgaJpZM4Lail1>
> .
>

lisavancasteren committed:
Feedback of Niels is implemented:
- button group
- functions replaced from controller to factorycontroller
- "you cannot afford another truck" added to language file

lisavancasteren commented:
The feedback has been implemented! I hope it is better?

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -32,6 +32,19 @@ class FactoryController extends OrderController
         );
 
+        //form submission, adds an extra truck
+        $ ("buy-truck").click(
```
Are you sure this selector works?

N-Wouda replied:
I think it needs to be `#buy-truck`, once the HTML has been updated.

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -105,6 +119,22 @@ class FactoryController extends OrderController
+    extraTruck (formValues)
+    {
+        if (this.validateTruck(truck)) {
+            GAME.model.config.money - GAME.model.config.costExtraTruck);
```
You'll need to assign this, or use `this.updateMoney(-GAME.model.config.costExtraTruck)`

N-Wouda commented:
I don't think you have everything yet, but I can see the direction it is going. Be sure to test your code while writing it! :)

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -105,6 +119,22 @@ class FactoryController extends OrderController
+    {
+        if (this.validateTruck(truck)) {
+            GAME.model.config.money - GAME.model.config.costExtraTruck);
+            GAME.model.config.maxSimultaneousOrders +1;
```
Same thing here, this update needs to be assigned (you may use `GAME.model.config.maxSimultaneousOrders++` too, that re-assigns in-place)

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -105,6 +119,22 @@ class FactoryController extends OrderController
+
+    validateTruck(truck)
+    {
+        if(costExtraTruck > GAME.model.config.money) {
```
`GAME.model.config.costExtraTruck`

lisavancasteren commented:
Updated! 
Tested while coding and a button appeared (yay!). Improved what happens to it when it is clicked, but now the whole game restarts when it is clicked....

lisavancasteren committed:
Feedback implemented and function behind button in factorycontroller adjusted.
A button appeared, but when it is clicked now, the game restarts...

N-Wouda commented on src/views/template/factory/factory.html:
```diff
@@ -32,16 +32,9 @@
-                {{#1}} Buy an extra truck{{/1}}
-            </button>
-        </form>
+        <form name="frequencyTrucks" action"">
```
This line needs to go. There is no need for a form here - it is just a button group.

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -32,18 +32,17 @@ class FactoryController extends OrderController
-                }
+        //form submission, adds an extra truck after cost validation
+        $("#buy-truck").submit(() => {
+            if(GAME.model.config.money>GAME.model.config.costExtraTruck){
```
Whitespaces around operators :)

N-Wouda replied:
So `if (GAME.model.etc. > GAME.model.etc.) {`.

N-Wouda commented on src/assets/language/nl.json:
```diff
@@ -37,6 +37,8 @@
   "There is no room for another order at this time!": "Er is op dit moment geen plek voor een nieuwe bestelling!",
   "There is insufficient space on the truck!": "Er is niet genoeg ruimte in de vrachtwagen!",
   "You cannot afford this!": "Je kunt dit niet betalen!",
+  "You cannot afford an extra truck!": "Je kunt geen extra vrachtwagen betalen!"
```
Comma at the end of this line.

N-Wouda commented on src/assets/language/nl.json:
```diff
@@ -37,6 +37,8 @@
   "There is insufficient space on the truck!": "Er is niet genoeg ruimte in de vrachtwagen!",
   "You cannot afford this!": "Je kunt dit niet betalen!",
+  "You cannot afford an extra truck!": "Je kunt geen extra vrachtwagen betalen!"
+  "You purchased an extra truck!": "Je hebt een extra vrachtwagen gekocht!"
```
And a comma here too

N-Wouda commented:
Very close haha!

lisavancasteren committed:
Small adjustments

lisavancasteren commented:
That's better! But when the button is clicked, nothing happens (no messages, no adjustments)

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -32,6 +32,18 @@ class FactoryController extends OrderController
         );
 
+        //form submission, adds an extra truck after cost validation
+        $("#buy-truck").submit(() => {
```
`click`, not `submit`. It is a button, not a form. Also, the comment above this line needs some updating :).

lisavancasteren committed:
Click instead of submit

lisavancasteren commented:
Aaah that's a difference
There is an issue with all of my buttons when I open the game, and I cannot update the LCfP/master to my branch anymore, think I accidentally deleted something somewhere :o

N-Wouda commented:
Have you taken care of the things I mentioned above? Especially the commas in the config files are important!

When you have, please make a commit. I cannot see what you are working on if you do not!

lisavancasteren commented:
Yes I took care of all comments, and just made a commit a few minutes ago

N-Wouda commented on src/controllers/core/controller.js:
```diff
@@ -109,9 +109,4 @@ class Controller
      *
      * @protected
      */
-    _updateMoney(amount)
```
We need this! Could you restore this function?

N-Wouda replied:
I wonder how you managed to remove this, but the result is at the very least, _amusing_. It happens to the best of us! 👍

N-Wouda commented:
I think I found it!

lisavancasteren commented:
It works!

lisavancasteren commented:
Thank you:)

lisavancasteren commented:
When I see the end result, that took me a looong time haha
Thanks for the feedback Niels:)

N-Wouda commented:
Practice makes perfect!

If you commit your changes, I'll give it a final review and walk you through the merge conflict somewhere tomorrow (then we can merge this and close the corresponding issue).

lisavancasteren commented:
All changes have been committed!

N-Wouda commented:
I still need you to restore that _updateMoney function!

lisavancasteren commented:
O I pasted it back in, that's not enough I think. Just turned off my
laptop, but will bring it with me tomorrow!

Op 11 jan. 2017 23:19 schreef "Niels Wouda" <notifications@github.com>:

> I still need you to restore that _updateMoney function!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LCfP/Serious-Game-for-Kids/pull/93#issuecomment-272013760>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV93kYDpin5x5x2GiWWiboWihWMs7ofdks5rRVVNgaJpZM4Lail1>
> .
>

N-Wouda commented:
You will have to commit those changes, but there is no real hurry. Tomorrow will be fine.

lisavancasteren committed:
Restore _updateMoney

lisavancasteren committed:
Restore _updateMoney

lisavancasteren committed:
Restore _updateMoney

lisavancasteren committed:
Restore _updateMoney

N-Wouda commented on src/controllers/core/controller.js:
```diff
@@ -102,16 +102,14 @@ class Controller
+        _updateMoney(amount)
+        {
+            GAME.model.config.money = GAME.model.config.money + parseFloat(amount);
+            $("#money").html(GAME.model.config.money.toFixed(2));
```
This is not correct. Could you carefully look at what it was like originally?

Else just copy this file https://raw.githubusercontent.com/LCfP/Serious-Game-for-Kids/master/src/controllers/core/controller.js and overwrite your entire local controller.js with it.

N-Wouda commented on src/controllers/factorycontroller.js:
```diff
@@ -32,6 +32,18 @@ class FactoryController extends OrderController
 
+        //adds an extra truck after cost validation
+        $("#buy-truck").click(() => {
+            if(GAME.model.config.money > GAME.model.config.costExtraTruck){
```
`if (...) {` whitespaces! I know this is a little pedantic, but a uniform style has its merits.

N-Wouda commented on src/views/template/factory/factory.html:
```diff
@@ -32,5 +32,8 @@
             </button>
         </form>
+        <div class="btn-group" role="group">
+            <button id="buy-truck" type="submit" class="btn btn-success pull-right margin-top-10">Buy truck</button>
```
`type="button"` seems more in order.

N-Wouda commented:
Okay I don't have much. These small remarks should suffice, and then it'll be ready to merge!

lisavancasteren committed:
Restore _updateMoney

N-Wouda commented on src/views/template/factory/factory.html:
```diff
@@ -32,5 +32,8 @@
             </button>
         </form>
+        <div class="btn-group" role="group">
+            <button id="buy-truck" type="click" class="btn btn-success pull-right margin-top-10">Buy truck</button>
```
`type="button"`!

N-Wouda commented on src/controllers/core/controller.js:
```diff
@@ -114,4 +114,4 @@ class Controller
         $("#money").html(GAME.model.config.money.toFixed(2));
     }
-}
+}
```
Could you add a single empty line to the end of the file here? After the bracket.

lisavancasteren committed:
Merge remote-tracking branch 'refs/remotes/LCfP/master' into frequency-trucks

# Conflicts:
#	src/controllers/factorycontroller.js

lisavancasteren committed:
Newline

N-Wouda commented:
This looks good to me.
